# Week 02 Assignment glucose level data

Welcome to week two of this course programming 1. You will learn about time related data wrangling with pandas and you will learn to visualize with bokeh. This week will be focussed around missing data. Concretely, you will preprocess the glucose json file using interpolation to impute in order to conduct visual analysis. Learning outcomes:

- load a json dataset 
- typecast the Pandas DataFrame to appropiate data types
- inspect the dataset for quality and metadata information
- add a column with interpolated data in Pandas DataFrame
- perform visual analysis

The assignment consists of 6 parts:

- [part 1: load the data](#0)
     - [Exercise 1.1](#ex-11)
- [part 2: prepare for inspection](#1)
     - [Exercise 2.1](#ex-21)
- [part 3: inspect the data](#2)
     - [Exercise 3.1](#ex-31)
- [part 4: interpolate the data](#3)
     - [Exercise 4.1](#ex-41)
- [part 5: visualize the data](#4)
     - [Exercise 5.1](#ex-51)
- [part 6: Challenge](#5)
     - [Exercise 6.1](#ex-61)

Part 1 and 5 are mandatory, part 6 is optional (bonus)
To pass the assingnment you need to a score of 60%. 


<a name='0'></a>
## Part 1: Load the data

Instructions: Load the json datafile `glucose.json` into a pandas dataframe. Check your dataframe with a `.head()` to compare with the expected outcome

<details>    
<summary>
    <font size="3" color="darkgreen"><b>Hints</b></font>
</summary>
<ul><li>json.load() method reads a file, pd.read_json converts it to a Pandas DataFrame</li>
    <li>when loading into a Pandas DataFrame use records orientation </li>
</ul>
</details>

<a name='ex-11'></a>
### Code your solution

In [5]:
import json
import pandas as pd

PATH = "/home/tijs/jupyter_notebooks/programming1/BFVM19PROG1/data/glucose.json"
f = open(PATH)
json_file = json.load(f)
df = pd.read_json(json_file)
df.head()

ID              time  recordtype glucose
0  2845.0  2019-04-25 00:08           1     109
1  2850.0  2019-04-25 00:50           1        
2  2877.0  2019-04-25 07:02           1     123
3  2881.0  2019-04-25 07:34           1     158
4  2886.0  2019-04-25 08:19           1

#### Expected outcome: 

<a name='1'></a>
## Part 2: Prepare the data

Check the datatypes of your dataframe. The `glucose` field should be an integer, the `time` field should have a datetime format. If the datatypes are different you should typecast them to the right format.
Make sure that your dataset is sorted by the time column


<details>    
<summary>
    <font size="3" color="darkgreen"><b>Hints</b></font>
</summary>
<ul><li>use astype() method or pandas.DataFrame.to_datetime() for instance</li>
    <li>make sure that the empty spaces are filled with NaN. Use errors='coerce'</li>
    <li>set_index(), sort_index() and reset_index() are helpful to sort on index</li>
</ul>
</details>

<a name='ex-21'></a>
### Code your solution

In [6]:
# CODE YOUR SOLUTION HERE
df.ID = pd.to_numeric(df.ID)
df.time = pd.to_datetime(df.time)
df.glucose = pd.to_numeric(df.glucose, errors='coerce')
print(df.dtypes)
df = df.set_index(['time']).sort_index()
df.head()

ID                   float64
time          datetime64[ns]
recordtype             int64
glucose              float64
dtype: object


ID  recordtype  glucose
time                                                  
2019-04-25 00:08:00  2.845000e+03           1    109.0
2019-04-25 00:14:00  1.614305e+19           0      NaN
2019-04-25 00:29:00  1.614305e+19           0      NaN
2019-04-25 00:44:00  1.614305e+19           0      NaN
2019-04-25 00:50:00  2.850000e+03           1      NaN

#### Expected outcome: 

<a name='2'></a>
## Part 3: Inspect the data

Now that we prepared the data we are going to inspect the data to get more familiar with the data. You are required to do the following

- inspect the percentage missing data for glucose
- what is the relationship between recordtype and glucose value?
- what is the relationship between ID and glucose value?

Code the solutions to your answers. Create meaningful overviews or statistics

<details>    
<summary>
    <font size="3" color="darkgreen"><b>Hints</b></font>
</summary>
<ul><li>In the week 01 assignment some functions were explained to inspect missing values</li>
    <li>In the week 01 assignment some functions were explained to groupby value</li>
</ul>
</details>

<a name='ex-31'></a>
### Code your solution

In [7]:
# inspect the percentage missing data for glucose
print(f"percentage missing data glucose: {df.glucose.isnull().sum()/len(df.glucose)}")
# what is the relationship between recordtype and glucose value?
print(df.groupby('recordtype').count())
print("If recordtype == 0 than glucose data is missing")
print(df.groupby('glucose', dropna=False).mean())
print("If ID is minimum numeric value (9223372036854775808) than glucose data is missing")

percentage missing data glucose: 0.6176470588235294
            ID  glucose
recordtype             
0           82        0
1           54       52
If recordtype == 0 than glucose data is missing
                   ID  recordtype
glucose                          
86.0     2.951000e+03     1.00000
91.0     2.940000e+03     1.00000
97.0     2.945500e+03     1.00000
105.0    2.974000e+03     1.00000
106.0    2.984000e+03     1.00000
107.0    2.979000e+03     1.00000
109.0    2.910500e+03     1.00000
110.0    3.026000e+03     1.00000
111.0    2.992667e+03     1.00000
114.0    3.004500e+03     1.00000
115.0    2.971000e+03     1.00000
116.0    3.031000e+03     1.00000
118.0    2.964500e+03     1.00000
120.0    2.983000e+03     1.00000
121.0    2.966500e+03     1.00000
122.0    2.999500e+03     1.00000
123.0    2.940000e+03     1.00000
124.0    3.002000e+03     1.00000
125.0    2.992000e+03     1.00000
126.0    3.028500e+03     1.00000
127.0    2.999000e+03     1.00000
128.0    3.020500e+03 

#### Expected outcome percentage missing data
0.6176470588235294

<a name='3'></a>
## Part 4: Interpolate the data

A lot of data is missing. We will use interpolation to fill the missing values. Create a new column with the interpolated data. 

<details>    
<summary>
    <font size="3" color="darkgreen"><b>Hints</b></font>
</summary>
<ul><li>use Pandas.DataFrame.interpolate() method</li>
</ul>
</details>

<a name='ex-41'></a>
### Code your solution

In [8]:
df['interpolated'] = df.glucose.interpolate()
df.head()

ID  recordtype  glucose  interpolated
time                                                                
2019-04-25 00:08:00  2.845000e+03           1    109.0    109.000000
2019-04-25 00:14:00  1.614305e+19           0      NaN    109.466667
2019-04-25 00:29:00  1.614305e+19           0      NaN    109.933333
2019-04-25 00:44:00  1.614305e+19           0      NaN    110.400000
2019-04-25 00:50:00  2.850000e+03           1      NaN    110.866667

#### Expected outcome

<a name='4'></a>
## Part 5: Plot the data

Create a plot with the original data and the interpolated data. An example of such a plot is given below.

<details>    
<summary>
    <font size="3" color="darkgreen"><b>Hints</b></font>
</summary>
<ul><li>figure(x_axis_type='datetime') automatically makes nices labels of the datetime data</li>
</ul>
</details>

<a name='ex-51'></a>
### Code your solution

In [14]:
from bokeh.io import output_notebook
from bokeh.plotting import figure, show
from bokeh.layouts import gridplot
from bokeh.io import output_notebook
from bokeh.plotting import ColumnDataSource
output_notebook()

Loading BokehJS ...

In [23]:
p = figure(title='Original vs interpolated Glucose',x_axis_type='datetime', background_fill_color="#fafafa")

###########################################
p.circle(df.index, df.glucose, color='red')
p.line(df.index, df.interpolated, color='blue')
###########################################

p.xaxis.axis_label = 'time'
p.yaxis.axis_label = 'signal'
p.grid.grid_line_color="white"

show(p)

<a name='6'></a>
## Part 6: Challenge

1. Can you improve the interpolation by choosing an other method?
2. Can you add an rolling mean line? 
2. Can you improve the plot by making it interactive?

<a name='ex-61'></a>
### Code your solution

In [18]:
#CODE YOUR SOLUTION HERE